### DISCUSSION TITLE
Is it feasible to use Snooze with CLOG?

### DISCUSSION THREAD

**Discussion by bhw-foss** [rabbibotton/clog#113] [Ideas]
Dear Dr. Botton,

CLOG has given a breath of new life into my web app, however I can't seem to get away entirely from REST/HTTP. I've chanced upon [snooze](https://github.com/joaotavora/snooze#rationale) as a URL router library I'd like to use but I can't seem to figure out how -- if it's possible at all -- to integrate it with CLOG. I've naively tried replacing `*app*` with `(snooze:make-clack-app)` in `clack:clackup` but it doesn't seem to work. I've no intention of asking to include snooze as a CLOG dependency, I think lispers put a premium on flexibility, so would just like to know if it's possible here to use snooze with CLOG. Or am I still too mired in the single duplex way of thinking? Either way, thanks again for CLOG!

Godspeed,

Hanshen

**Comment by rabbibotton** [comment]
It should be no issue to use along side clog. (I haven't tried this so let me know if had to make any changes :)

clog-connects::*app* is clog's middleware for clack after initialization.

You need to wrap the snooze middleware around that:  (setf clog-connection::*app* (funcall (snooze:make-clack-app) clog-connection::*app*))

I don't think you can change clack middleware while running so for the moment do the following _after_ initializing your CLOG app:

(clack:stop clog-connection::*client-handler*)
(setf clog-connection::*app* (funcall (snooze:make-clack-app) clog-connection::*app*))
(setf clog-connection::*client-handler* (clack:clackup clog-connection::*app* :address "0.0.0.0" :port 8080))

Let me know how it goes, if works I'll add something to allow adding middleware's before init of clog.

**Reply by bhw-foss** [reply]
Running `(setf clog-connection::*app* (funcall (snooze:make-clack-app) clog-connection::*app*))` and sbcl points out to me,
```
The value
  #<FUNCTION (LAMBDA (CLOG-CONNECTION::ENV)
               :IN
               CLOG-CONNECTION:INITIALIZE) {10083B15DB}>

is not of type
  LIST
when binding SB-IMPL::PLACE
   [Condition of type TYPE-ERROR]
```
but it's no biggie, I ended up going with,

```
(defclass snooze-acceptor (hunchensocket:websocket-acceptor
                           hunchentoot:easy-acceptor) ())

...

(setf hunchentoot:*dispatch-table*
      (list (hunchentoot:create-folder-dispatcher-and-handler
             "/public/" (asdf:system-relative-pathname :your-project "assets/"))
            (snooze:make-hunchentoot-app '((*home-resource* . homepage)))))

(setf hunchensocket:*websocket-dispatch-table*
      (list 'find-room))

(hunchentoot:start (make-instance 'snooze-acceptor))
```

though I'm sure there's a way of getting it to run on the same port with clack, I was just more comfortable with hunchentoot. Thanks for the pointers!

**Comment by rabbibotton** [comment]
BTW all of that was if you wanted run snooze on same port - there is no issue at all if running your REST apis on say http://127.0.0.1:9003 and then CLOG is running on http://127.0.0.1:8080 

In a case like that (something I would do to start with) just run your clog app and your snooze app (clack:clackup (snooze:make-clack-app) :port 9003)

**Comment by bhw-foss** [comment]
@rabbibotton Hey Dr. Botton, I'm back and trying to make CLOG and snooze work together. I have some questions regarding tutorial 11,

1. Is it possible for the bootfile to be dynamically generated? What I mean is I use snooze routing to generate HTML (with generated <scripts> boot.js and jquery.min.js) on port 9003 and then `clog:attach-as-child` to perform CLOG magic. Does `:boot-file` have to be static?
2. For `clog:set-on-new-window` the `:path` docstring says paths must begin with a forward slash "/". So if I wanted to enable CLOG on the aforementioned snooze generated HTML `(set-on-new-window 'foobar :path "http://127.0.0.1:9003/")` would be the right idea?
3. Happily since the last time I looked at the problem, Shaka has added `:lack-middleware-list`. When I try to use snooze then as clack middleware (source is https://github.com/joaotavora/snooze/pull/23) I get the same error. Here is my reproduction snippet.

```
(defpackage #:clog-tut-1
  (:use #:cl #:clog #:snooze)
  (:export start-tutorial))

(in-package :clog-tut-1)

(defun on-new-window (body)
  "On-new-window handler."
  (let ((hello-element (create-child body "<h1>Hello World! (click me!)</h1>")))
    (set-on-click hello-element
                  (lambda (obj)
                    (declare (ignore obj))
                    (setf (color hello-element) :green)))))

;;; http://127.0.0.1:8080/test/123
(defroute test (:get "text/html" (foobar integer))
  (format nil "<h1>Hello World from Snooze, ~a </h1>" foobar))

;;; http://127.0.0.1:8080/test/symbol
(defroute test (:get "text/html" (foobar symbol))
  (format nil "<h1>Hello World from Snooze, ~a </h1>" foobar))

(defun make-clack-middleware (&optional bindings)
  "Install Snooze in a Clack stack as middleware.
A wrapper around make-clack-app that passes requests down the stack when a 404
is returned by Snooze."
  (lambda (app)
    (let ((snooze-app (make-clack-app bindings)))
      (lambda (env)
        (let ((results (funcall snooze-app env)))
          (if (eq 404 (car results))
              (funcall app env)
              results))))))

(defun start-tutorial ()
  "Start turtorial."
  (initialize #'on-new-window :lack-middleware-list '(make-clack-middleware))
  (open-browser))
```

and here is the error I get,

```
The value
  #<FUNCTION (LAMBDA (CLOG-CONNECTION::ENV)
               :IN
               CLOG-CONNECTION:INITIALIZE) {1002AA0CEB}>

is not of type
  LIST
   [Condition of type TYPE-ERROR]

Restarts:
 0: [ABORT] abort thread (#<THREAD "hunchentoot-worker-127.0.0.1:40966" RUNNING {1005B7B653}>)

Backtrace:
 0: ((LAMBDA (SNOOZE::ENV) :IN MAKE-CLACK-APP) #<FUNCTION (FLET CLACK.HANDLER.HUNCHENTOOT::HANDLE-NORMAL-RESPONSE :IN CLACK.HANDLER.HUNCHENTOOT::HANDLE-RESPONSE) {53D6547B}>)
 1: ((LAMBDA (ENV) :IN MAKE-CLACK-MIDDLEWARE) #<FUNCTION (FLET CLACK.HANDLER.HUNCHENTOOT::HANDLE-NORMAL-RESPONSE :IN CLACK.HANDLER.HUNCHENTOOT::HANDLE-RESPONSE) {53D6547B}>)
 2: (CLACK.HANDLER.HUNCHENTOOT::HANDLE-RESPONSE #<FUNCTION (LAMBDA (ENV) :IN MAKE-CLACK-MIDDLEWARE) {10069DE9CB}>)
 3: ((:METHOD HUNCHENTOOT:ACCEPTOR-DISPATCH-REQUEST (CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR T)) #<CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR (host 0.0.0.0, port 8080)> #<HUNCHENTOOT:REQUEST {1006715F33}..
 4: ((:METHOD HUNCHENTOOT:HANDLE-REQUEST (HUNCHENTOOT:ACCEPTOR HUNCHENTOOT:REQUEST)) #<CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR (host 0.0.0.0, port 8080)> #<HUNCHENTOOT:REQUEST {1006715F33}>) [fast-metho..
 5: ((:METHOD HUNCHENTOOT:PROCESS-REQUEST (T)) #<HUNCHENTOOT:REQUEST {1006715F33}>) [fast-method]
 6: (HUNCHENTOOT::DO-WITH-ACCEPTOR-REQUEST-COUNT-INCREMENTED #<CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR (host 0.0.0.0, port 8080)> #<FUNCTION (LAMBDA NIL :IN HUNCHENTOOT:PROCESS-CONNECTION) {10064683DB}>..
 7: ((:METHOD HUNCHENTOOT:PROCESS-CONNECTION (HUNCHENTOOT:ACCEPTOR T)) #<CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR (host 0.0.0.0, port 8080)> #<USOCKET:STREAM-USOCKET {1005B794E3}>) [fast-method]
 8: ((:METHOD HUNCHENTOOT:PROCESS-CONNECTION :AROUND (HUNCHENTOOT:ACCEPTOR T)) #<CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR (host 0.0.0.0, port 8080)> #<USOCKET:STREAM-USOCKET {1005B794E3}>) [fast-method]
 9: ((:METHOD HUNCHENTOOT:PROCESS-CONNECTION :AROUND (CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR T)) #<CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR (host 0.0.0.0, port 8080)> #<USOCKET:STREAM-USOCKET {1005B79..
10: ((:METHOD HUNCHENTOOT::HANDLE-INCOMING-CONNECTION% (HUNCHENTOOT:ONE-THREAD-PER-CONNECTION-TASKMASTER T)) #<HUNCHENTOOT:ONE-THREAD-PER-CONNECTION-TASKMASTER {100508AC03}> #<USOCKET:STREAM-USOCKET {1005..
11: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
12: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
13: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::RUN))
14: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
15: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::RUN))
16: (SB-THREAD::RUN)
17: ("foreign function: call_into_lisp_")
18: ("foreign function: funcall1")
```
Thank you very much for your time.

**Reply by rabbibotton** [reply]
Form docstring on Initialize -

boot-function if set is called with the
url and the contents of boot-file and its return value replaces the contents
sent to the brower, this allows adding content for search engine optimization,
see tutorial 12 for an example.

**Reply by rabbibotton** [reply]
<<For clog:set-on-new-window the :path docstring says paths must begin with a forward slash "/". So if I wanted to enable CLOG on the aforementioned snooze generated HTML (set-on-new-window 'foobar :path "http://127.0.0.1:9003/") would be the right idea?>>
(you probably got emails with my original response, sorry I deleted it to move the reply and messed up a copy and paste)

You would not do this as you do not need a CLOG on-new-window handler for snooze. Instead as I mentioned Feb 18:
just run your clog app defaulting to its port say 8080  and then your snooze app (clack:clackup (snooze:make-clack-app) :port 9003) and they run in parallel

**Reply by bhw-foss** [reply]
I apologize, I feel like I am missing something very obvious. I have CLOG and snooze running on different ports, I do have a minimal example:

```
(defpackage #:clog-tut-1
  (:use #:cl #:clog #:snooze)
  (:export start-tutorial))

(in-package :clog-tut-1)

(defun on-main (body)
  "On-new-window handler."
  (let ((title-element (attach-as-child body "title")))
    (set-on-click title-element
                  (lambda (obj)
                    (declare (ignore obj))
                    (setf (color title-element) :green)))))

;; http://127.0.0.1:9003/test/hello
;; (test (make-instance 'snooze-verbs:get) (make-instance 'snooze-types:text/html) 'hello)
(defroute test (:get "text/html" foobar)
  (format nil "<script src=\"/jquery.min.js\" type=\"text/javascript\"></script>
               <script src=\"/boot.js\" type=\"text/javascript\"></script>
               <h1 id=\"title\">Hello World from Snooze, ~a </h1>"
          foobar))

(defun start-tutorial ()
  "Start turtorial."
  (initialize #'on-main
              :port 8080
              :long-poll-first t
              :extended-routing t)
  (setf hunchentoot:*dispatch-table*
        (list (hunchentoot:create-static-file-dispatcher-and-handler
               "/jquery.min.js"
               (asdf:system-relative-pathname "clog" "static-files/js/jquery.min.js"))
              (hunchentoot:create-static-file-dispatcher-and-handler
               "/boot.js"
               (asdf:system-relative-pathname "clog" "static-files/js/boot.js"))
              (snooze:make-hunchentoot-app)))
  (hunchentoot:start (make-instance 'hunchentoot:easy-acceptor :port 9003)))

;; Then http://127.0.0.1:9003/test/hello and clicking <h1> will turn the text green.
;; Or do I have to go to http://127.0.0.1:8080/ and use :boot-function to generate the html?
```

**Reply by rabbibotton** [reply]
So snooze is for calls to do Ajax calls that return data not returning pages to a browser. You for example wouldn't go to a snooze server to return a web page.

So snooze you would use for example to provide 3rd party access to data on your system, and say use CLOG for web pages and controls panels.

Because CLOG is a much more efficient way to communicate to a server you wouldn't call from a web page to your own Ajax snooze server. You would just internal use the data and output with CLOG.

So maybe give me a practical example of how you imagined using this and I can help better and maybe even show you an example.

**Reply by bhw-foss** [reply]
Okay, I guess that makes sense. Thanks Dr. Botton.

**Reply by rabbibotton** [reply]
how did it work out?

**Reply by bhw-foss** [reply]
Happy new years Rabbi! https://github.com/rabbibotton/clog/discussions/310 is what I wanted but didn't know how to articulate. Clog is real neat.
